ext/spl/spl_array.c

```diff
@@ -477,11 +480,23 @@ static zval *spl_array_read_dimension(zend_object *object, zval *offset, int typ
 	return spl_array_read_dimension_ex(1, object, offset, type, rv);
 } /* }}} */
 
+static uint32_t spl_array_set_refcount(spl_array_object *intern, HashTable *ht, uint32_t refcount) /* {{{ */
+{
+	uint32_t old_refcount = 0;
+	if (intern->is_child) {
+		old_refcount = GC_REFCOUNT(ht);
+		GC_SET_REFCOUNT(ht, refcount);
+	}
+
+	return old_refcount;
+} /* }}} */
```

Comment:
I'm really confused by what's the point of this function staring at the usage.

Is this to detect if an SPLArray is a child and is refcounted?

Reply:
```c
ZVAL_ARR(&intern->array, zend_array_dup(Z_ARR_P(array)));
if (intern->is_child) {
    Z_TRY_DELREF_P(&intern->bucket->val);
	intern->bucket->val = intern->array;
	Z_TRY_ADDREF_P(&intern->array);
}
```
Because I added the `refcount` manually here and the assertion `HT_ASSERT_RC1(ht)` will failed.
```c
static zend_always_inline zval *_zend_hash_index_add_or_update_i(HashTable *ht, zend_ulong h, zval *pData, uint32_t flag)
{
    ........
	IS_CONSISTENT(ht);
	HT_ASSERT_RC1(ht);
    ........

#define HT_ASSERT_RC1(ht) HT_ASSERT(ht, GC_REFCOUNT(ht) == 1)
```

So I have to set the `refcount` to 1 when `intern->is_child` is true to make assertion success and restore the  `refcount` to the original value after modifying the `array`.